substrate/client/network/statement/src/lib.rs

```diff
@@ -470,9 +516,55 @@ where
 
 			propagated_statements += to_send.len();
 
-			if !to_send.is_empty() {
-				log::trace!(target: LOG_TARGET, "Sending {} statements to {}", to_send.len(), who);
-				self.notification_service.send_sync_notification(who, to_send.encode());
+			let mut offset = 0;
+			while offset < to_send.len() {
+				// Try to send as many statements as possible in one notification
+				let chunk_size = to_send.len() - offset;
+				let mut current_end = to_send.len();
+
+				loop {
+					let chunk = &to_send[offset..current_end];
+					let encoded_size = chunk.encoded_size();
+
+					// If chunk fits, send it
+					if encoded_size <= MAX_STATEMENT_NOTIFICATION_SIZE as usize {
+						log::trace!(
+							target: LOG_TARGET,
+							"Sending {} statements ({} KB) to {}",
+							chunk.len(),
+							encoded_size / 1024,
+							who
+						);
+						self.notification_service.send_sync_notification(who, chunk.encode());
+						offset = current_end;
+						if let Some(ref metrics) = self.metrics {
+							metrics.propagated_statements_chunks.observe(chunk.len() as f64);
+						}
+						break;
+					}
+
+					// Size exceeded - split the chunk
+					let split_factor =
+						(encoded.len() / MAX_STATEMENT_NOTIFICATION_SIZE as usize) + 1;
+					let new_chunk_size = (current_end - offset) / split_factor;
+
+					// Single statement is too large
+					if new_chunk_size == 0 {
+						log::warn!(
+							target: LOG_TARGET,
+							"Statement too large ({} KB), skipping",
```

Comment:
It would be nice to exactly skip the one statement that is problematic and not maybe more.

Reply:
You're right! 😱